### DISCUSSION TITLE
The activity function 'CheckHttpChallenge' failed - Challenge file not created in correct location?

### DISCUSSION THREAD

**Discussion by tigreye007** [shibayan/appservice-acmebot#276] [Q&A] [answered]
We are attempting to implement this for the first time, and when attempting to issue a cert, we receive the following message:

_Orchestrator function 'AddCertificate_Orchestrator' failed: The orchestrator function 'IssueCertificate' failed: "The activity function 'CheckHttpChallenge' failed: "[redacted url] is InternalServerError status code.". See the function execution logs for additional details.". See the function execution logs for additional details_

When I FTP into the app service (Windows/.NET application), I see that the file was created at /site/.well-known, however our previous LetsEncrypt web job would create files at /site/wwwroot/.well-known, which makes sense as our site root is /site/wwwroot.

I see in the readme that we should try setting inheritInChildApplications="false" in the main application's web.config (/site/wwwroot/web.config), however that is meant for child applications (or anything under wwwroot in our case).  Is there some sort of configuration we're missing that would prevent the function from creating the file in the correct location?

**Comment by shibayan** [comment, marked as answer]
Since it strongly depends on the Web.config used by the application, you need to customize `site/.well-known/web.config` by referring to the following issue.

#243
#251

**Reply by tigreye007** [reply]
I have updated the web.config in /site/.well-known as follows, but it still errors out. It looks like that web.config was overwritten - what else might I be missing? Thanks!

`<?xml version="1.0" encoding="utf-8"?> <configuration> <system.webServer> <handlers> <clear /> <add name="StaticFile" path="*" verb="*" modules="StaticFileModule" resourceType="Either" requireAccess="Read" /> </handlers> <staticContent> <remove fileExtension="." /> <mimeMap fileExtension="." mimeType="text/plain" /> </staticContent> <rewrite> <rules> <clear /> </rules> </rewrite> </system.webServer> <system.web> <authorization> <allow users="*"/> </authorization> </system.web> <location path="." inheritInChildApplications="false"></location> </configuration>`

**Reply by tigreye007** [reply]
I have also tried adding the following to the application web.config at /site/wwwroot, and still getting the same error:
```

  <location path="." inheritInChildApplications="false">
  </location>
```

**Reply by shibayan** [reply]
In order to use a custom web.config, you need to put a file called `configured` in the same directory as web.config.
https://github.com/shibayan/appservice-acmebot#in-case-you-want-to-use-your-own-webconfig

**Reply by tigreye007** [reply]
Thanks @shibayan - adding an empty "configured" file worked for preventing the /site/.well-known/web.config from being overridden, however we are still getting the same CheckHttpChallenge exception.  Maybe I'm not understanding how files would be served from /site/.well-known when the app is served from /site/wwwroot/.  Our old web job would put the challenge file in /site/wwwroot/.well-known, and would succeed, while this azure function places the challenge file at /site/.well-known.  Anything else you can think of that I might be missing?

**Reply by shibayan** [reply]
You may want to check if you can access the files under .well-known in that state. Further support is not possible.

**Reply by tigreye007** [reply]
Thank you again for your reply.  I think I may have figured out what's going on - the code appears to attempt to create a Virtual Application for /.well-known, however when I look at my app service config, I do not see the virtual application:

![virtual](https://user-images.githubusercontent.com/1466164/107268110-564df200-6a0d-11eb-9f87-e544a4f4eea5.PNG)

Would there be anything in the logs if this failed?  Where can I find the logs?

**Reply by shibayan** [reply]
The ".well-known" virtual application will be removed upon completion regardless of the success or failure of the process. You will need to add them manually.

**Reply by tigreye007** [reply]
Thanks for all your help.  With a little bit of debugging, I was able to determine that all I needed was the following web.config (and empty **configured** file in /site/.well-known.  This should work for anyone using Umbraco 8:
```

<?xml version="1.0" encoding="utf-8"?>
<configuration>
  <system.webServer>
    <handlers>
      <clear />
      <add name="StaticFile" path="*" verb="*" modules="StaticFileModule" resourceType="Either" requireAccess="Read" />
    </handlers>
    <staticContent>
      <remove fileExtension="." />
      <mimeMap fileExtension="." mimeType="text/plain" />
    </staticContent>
    <rewrite>
      <rules>
        <clear />
      </rules>
    </rewrite>
    <modules>
        <remove name="UmbracoModule" />
        <remove name="ClientDependencyModule" />
        <remove name="ImageProcessorModule" />
        <remove name="RedirectsModule" />
        <remove name="RoleManager" />
    </modules>
  </system.webServer>
  <system.web>
    <authorization>
      <allow users="*"/>
    </authorization>
    <httpModules>
        <remove name="UmbracoModule" />
        <remove name="ClientDependencyModule" />
        <remove name="ImageProcessorModule" />
    </httpModules>
  </system.web>
</configuration>
```

**Comment by christiantricarico** [comment]
I'm having the same problem ... how did you fix it?

**Reply by tigreye007** [reply]
Please see above thread:  https://github.com/shibayan/appservice-acmebot/discussions/276#discussioncomment-353618

Depending on your application, you may need to remove certain modules from the virtual path by updating the web.config in the .well-known folder.  You can test your web.config implementation by manually adding a virtual application to your app service in Azure with the following settings:

1. Virtual path: /.well-known
2. Physical Path: site\.well-known
3. Uncheck "Directory"
4. Uncheck "Preload enabled"
5. Click OK and then Save the configuration updates.

Once that's in place, you can manually visit your http://site.com/.well-known/acme-challenge/[code] to see if your updates worked.  You can also turn custom errors off in the web.config to help get an idea of what other modules might be getting in the way.

**Reply by christiantricarico** [reply]
Thank you for your answer.
My website is a Umbraco V7 app.
Now I created the virtual app in this way:
![image](https://user-images.githubusercontent.com/8735590/118692981-ea003d00-b80a-11eb-94d1-b4a2bcd69639.png)

Updated the web.config in .well-known as suggested by you and created an empty "configured" file.
When I visit the code in acme-challenge I get this error message:
![image](https://user-images.githubusercontent.com/8735590/118693270-351a5000-b80b-11eb-9962-7b52b08896e3.png)

Does it seem the well-known app is inheriting some extra setting from the root web.config?

**Reply by tigreye007** [reply]
My example web.config above was for Umbraco 8.  I would look through the web.config for your Umbraco 7 project to see what other modules may need removal.

**Reply by christiantricarico** [reply]
After clearing the the assemblies section in child web.config I can finally visit the /.well-known/acme-challenge/[code] page without errors.

But now the error generated by the issuing function is different and I can't understand what is going wrong:
![image](https://user-images.githubusercontent.com/8735590/118776218-09897b00-b888-11eb-8f49-e14c63d0b2d1.png)

I can't find anything googling CheckIsReady. Any idea?
